posthog/models/cohort.py

```diff
@@ -102,39 +110,39 @@ def get_analytics_metadata(self):
             "deleted": self.deleted,
         }
 
-    def calculate_people(self):
+    def calculate_people(self, batch_size=50000, pg_batch_size=1000):
         if self.is_static:
             return
         try:
-            persons_query = self._clickhouse_persons_query()
-
-            try:
-                sql, params = persons_query.distinct("pk").only("pk").query.sql_with_params()
-            except EmptyResultSet:
-                query = DELETE_QUERY.format(cohort_id=self.pk)
-                params = {}
-            else:
-                query = f"""
-                    {DELETE_QUERY};
-                    {UPDATE_QUERY};
-                """.format(
-                    cohort_id=self.pk,
-                    values_query=sql.replace('FROM "posthog_person"', f', {self.pk} FROM "posthog_person"', 1,),
-                )
-
-            cursor = connection.cursor()
+            start_time = time.time()
+            # Get or wait to update cohort version
             with transaction.atomic():
```

Comment:
Shouldn't we have this section at the end? Otherwise we could reference a version that is not yet available

Reply:
The sequence I would go for is:

 1. generate some version identifier, could just be some uuid, a postgresql sequence number
 2. create CohortPeople with this version associated
 3. update the Cohort with the new version, but do something like an UPDATE cohort SET version=newversion WHERE version = oldversion although we could probably so without the version = oldversion check, I'm not sure how much we care about avoiding updating again but probably wont do any harm.

If we want to make it easy to find orphans we can use an increasing version and look for people with older versions. wdyt

Reply:
Oh i was assuming they would be queue'd in order but that might not be true hmm. I don't think before or after updating will solve the race condition. If there are multiple updates to be processed, how do we guarantee the latest one that the user asked for will be the last update

Reply:
I'm not sure how workers/jobs are queued/enqueued but can we enforce to run a single concurrent worker process per cohort?

Reply:
I think we can get rid of the time stamp and rather use an explicit incremental version to avoid relying on time stamps

Reply:
Because of the ui changes?

Reply:
We rely on Postgres for generation of a strictly increasing token, we pass this explicitly to other process e.g. passing the celery task